Micronaut Data JDBC Criteria API multiselect fields not working

### Expected Behavior

Retrieve only the fields requested in the CriteriaQuery multiselect.

```java
    public static <T> QuerySpecification<T> fieldSelector() {
        return (root, query, cb) -> {
            query.multiselect(root.get("title").alias("title"), root.get("author").alias("author"));
            return query.getRestriction();
        };
    }
```

This works fine with Micronaut Data JPA

### Actual Behaviour


```java
Error reading object for name [id] from result set: Column "id" not found [42122-232]
io.micronaut.data.exceptions.DataAccessException: Error reading object for name [id] from result set: Column "id" not found [42122-232]
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.exceptionForColumn(ColumnNameResultSetReader.java:263)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readInt(ColumnNameResultSetReader.java:169)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readInt(ColumnNameResultSetReader.java:43)
	at io.micronaut.data.runtime.mapper.ResultReader.readDynamic(ResultReader.java:107)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readDynamic(ColumnNameResultSetReader.java:69)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readDynamic(ColumnNameResultSetReader.java:43)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.readProperty(SqlResultEntityTypeMapper.java:702)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.readEntityId(SqlResultEntityTypeMapper.java:728)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.readEntity(SqlResultEntityTypeMapper.java:474)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper$4.processRow(SqlResultEntityTypeMapper.java:370)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:397)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:386)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:378)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$findAll$5(DefaultJdbcRepositoryOperations.java:530)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$executeRead$23(DefaultJdbcRepositoryOperations.java:770)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.withExistingConnectionInternal(AbstractConnectionOperations.java:128)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.execute(AbstractConnectionOperations.java:92)
	at io.micronaut.data.connection.ConnectionOperations.executeRead(ConnectionOperations.java:71)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.executeRead(DefaultJdbcRepositoryOperations.java:767)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:530)
	at io.micronaut.data.runtime.intercept.criteria.AbstractSpecificationInterceptor.findAll(AbstractSpecificationInterceptor.java:156)
	at io.micronaut.data.runtime.intercept.criteria.FindAllSpecificationInterceptor.intercept(FindAllSpecificationInterceptor.java:46)
	at io.micronaut.data.runtime.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:83)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:143)
	at com.example.data.repository.BookRepository$Intercepted.findAll(Unknown Source)
	at com.example.BookRepositoryTest.testFieldSelector(BookRepositoryTest.java:45)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension$2.proceed(MicronautJunit5Extension.java:142)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:162)
	at io.micronaut.test.extensions.AbstractMicronautExtension$3.proceed(AbstractMicronautExtension.java:174)
	at io.micronaut.test.context.TestMethodInterceptor.interceptTest(TestMethodInterceptor.java:46)
	at io.micronaut.transaction.test.DefaultTestTransactionExecutionListener.lambda$interceptTest$0(DefaultTestTransactionExecutionListener.java:93)
	at io.micronaut.transaction.support.AbstractPropagatedStatusTransactionOperations.lambda$execute$2(AbstractPropagatedStatusTransactionOperations.java:68)
	at io.micronaut.transaction.TransactionCallback.apply(TransactionCallback.java:37)
	at io.micronaut.transaction.support.AbstractTransactionOperations.executeTransactional(AbstractTransactionOperations.java:333)
	at io.micronaut.transaction.support.AbstractTransactionOperations.executeWithNewTransaction(AbstractTransactionOperations.java:318)
	at io.micronaut.transaction.support.AbstractTransactionOperations.executeNew(AbstractTransactionOperations.java:235)
	at io.micronaut.transaction.support.AbstractTransactionOperations.doExecute(AbstractTransactionOperations.java:137)
	at io.micronaut.transaction.support.AbstractTransactionOperations.lambda$doExecute$0(AbstractTransactionOperations.java:122)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.executeWithNewConnection(AbstractConnectionOperations.java:143)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.execute(AbstractConnectionOperations.java:90)
	at io.micronaut.transaction.support.AbstractTransactionOperations.doExecute(AbstractTransactionOperations.java:120)
	at io.micronaut.transaction.support.AbstractPropagatedStatusTransactionOperations.execute(AbstractPropagatedStatusTransactionOperations.java:65)
	at io.micronaut.transaction.test.DefaultTestTransactionExecutionListener.interceptTest(DefaultTestTransactionExecutionListener.java:91)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:166)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptTest(AbstractMicronautExtension.java:119)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.interceptTestMethod(MicronautJunit5Extension.java:129)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Column "id" not found [42122-232]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:514)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
	at org.h2.message.DbException.get(DbException.java:223)
	at org.h2.message.DbException.get(DbException.java:199)
	at org.h2.jdbc.JdbcResultSet.getColumnIndex(JdbcResultSet.java:3518)
	at org.h2.jdbc.JdbcResultSet.getInt(JdbcResultSet.java:337)
	at com.zaxxer.hikari.pool.HikariProxyResultSet.getInt(HikariProxyResultSet.java)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readInt(ColumnNameResultSetReader.java:167)
	... 47 more


Column "id" not found [42122-232]
org.h2.jdbc.JdbcSQLSyntaxErrorException: Column "id" not found [42122-232]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:514)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
	at org.h2.message.DbException.get(DbException.java:223)
	at org.h2.message.DbException.get(DbException.java:199)
	at org.h2.jdbc.JdbcResultSet.getColumnIndex(JdbcResultSet.java:3518)
	at org.h2.jdbc.JdbcResultSet.getInt(JdbcResultSet.java:337)
	at com.zaxxer.hikari.pool.HikariProxyResultSet.getInt(HikariProxyResultSet.java)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readInt(ColumnNameResultSetReader.java:167)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readInt(ColumnNameResultSetReader.java:43)
	at io.micronaut.data.runtime.mapper.ResultReader.readDynamic(ResultReader.java:107)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readDynamic(ColumnNameResultSetReader.java:69)
	at io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader.readDynamic(ColumnNameResultSetReader.java:43)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.readProperty(SqlResultEntityTypeMapper.java:702)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.readEntityId(SqlResultEntityTypeMapper.java:728)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper.readEntity(SqlResultEntityTypeMapper.java:474)
	at io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper$4.processRow(SqlResultEntityTypeMapper.java:370)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:397)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:386)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:378)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$findAll$5(DefaultJdbcRepositoryOperations.java:530)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$executeRead$23(DefaultJdbcRepositoryOperations.java:770)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.withExistingConnectionInternal(AbstractConnectionOperations.java:128)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.execute(AbstractConnectionOperations.java:92)
	at io.micronaut.data.connection.ConnectionOperations.executeRead(ConnectionOperations.java:71)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.executeRead(DefaultJdbcRepositoryOperations.java:767)
	at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.findAll(DefaultJdbcRepositoryOperations.java:530)
	at io.micronaut.data.runtime.intercept.criteria.AbstractSpecificationInterceptor.findAll(AbstractSpecificationInterceptor.java:156)
	at io.micronaut.data.runtime.intercept.criteria.FindAllSpecificationInterceptor.intercept(FindAllSpecificationInterceptor.java:46)
	at io.micronaut.data.runtime.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:83)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:143)
	at com.example.data.repository.BookRepository$Intercepted.findAll(Unknown Source)
	at com.example.BookRepositoryTest.testFieldSelector(BookRepositoryTest.java:45)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension$2.proceed(MicronautJunit5Extension.java:142)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:162)
	at io.micronaut.test.extensions.AbstractMicronautExtension$3.proceed(AbstractMicronautExtension.java:174)
	at io.micronaut.test.context.TestMethodInterceptor.interceptTest(TestMethodInterceptor.java:46)
	at io.micronaut.transaction.test.DefaultTestTransactionExecutionListener.lambda$interceptTest$0(DefaultTestTransactionExecutionListener.java:93)
	at io.micronaut.transaction.support.AbstractPropagatedStatusTransactionOperations.lambda$execute$2(AbstractPropagatedStatusTransactionOperations.java:68)
	at io.micronaut.transaction.TransactionCallback.apply(TransactionCallback.java:37)
	at io.micronaut.transaction.support.AbstractTransactionOperations.executeTransactional(AbstractTransactionOperations.java:333)
	at io.micronaut.transaction.support.AbstractTransactionOperations.executeWithNewTransaction(AbstractTransactionOperations.java:318)
	at io.micronaut.transaction.support.AbstractTransactionOperations.executeNew(AbstractTransactionOperations.java:235)
	at io.micronaut.transaction.support.AbstractTransactionOperations.doExecute(AbstractTransactionOperations.java:137)
	at io.micronaut.transaction.support.AbstractTransactionOperations.lambda$doExecute$0(AbstractTransactionOperations.java:122)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.executeWithNewConnection(AbstractConnectionOperations.java:143)
	at io.micronaut.data.connection.support.AbstractConnectionOperations.execute(AbstractConnectionOperations.java:90)
	at io.micronaut.transaction.support.AbstractTransactionOperations.doExecute(AbstractTransactionOperations.java:120)
	at io.micronaut.transaction.support.AbstractPropagatedStatusTransactionOperations.execute(AbstractPropagatedStatusTransactionOperations.java:65)
	at io.micronaut.transaction.test.DefaultTestTransactionExecutionListener.interceptTest(DefaultTestTransactionExecutionListener.java:91)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:166)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptTest(AbstractMicronautExtension.java:119)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.interceptTestMethod(MicronautJunit5Extension.java:129)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)

```

### Steps To Reproduce

1. Download the example application
2. Use the "jdbc" branch `git checkout jdbc`
3. Run the test `./gradlew test --tests "com.example.BookRepositoryTest.testFieldSelector"`

The same implementation with Micronaut Data JPA can be found in the branch "jpa" `git checkout jpa`


### Environment Information

- Operating System: MacOS Sonoma
- JDK: 21

### Example Application

https://github.com/ralmeida7/criteria-builder-bugs

### Version

4.6.0

--- a/data-connection/src/main/java/io/micronaut/data/connection/support/AbstractReactorConnectionOperations.java
+++ b/data-connection/src/main/java/io/micronaut/data/connection/support/AbstractReactorConnectionOperations.java
@@ -99,13 +99,13 @@ public abstract class AbstractReactorConnectionOperations<C> implements ReactorC
     }
 
     private <T> Flux<T> existingConnectionFlux(ConnectionDefinition definition, Function<ConnectionStatus<C>, Flux<T>> callback, C clientSession) {
-        return callback.apply(new DefaultReactiveConnectionStatus<>(clientSession, definition, false));
+        return applyCallbackFlux(callback, new DefaultReactiveConnectionStatus<>(clientSession, definition, false));
     }
 
     private <T> Flux<T> openConnectionFlux(ConnectionDefinition definition, Function<ConnectionStatus<C>, Flux<T>> callback) {
         return Flux.usingWhen(
             Mono.from(openConnection(definition)).map(connection -> new DefaultReactiveConnectionStatus<>(connection, definition, true)),
-            connectionStatus -> callback.apply(connectionStatus).contextWrite(ctx -> addClientSession(ctx, connectionStatus)),
+            connectionStatus -> applyCallbackFlux(callback, connectionStatus).contextWrite(ctx -> addClientSession(ctx, connectionStatus)),
             connectionStatus -> connectionStatus.onComplete(() -> closeConnection(connectionStatus.getConnection(), definition)),
             (connectionStatus, throwable) -> connectionStatus.onError(throwable, () -> closeConnection(connectionStatus.getConnection(), definition)),
             connectionStatus -> connectionStatus.onCancel(() -> closeConnection(connectionStatus.getConnection(), definition))
@@ -134,13 +134,13 @@ public abstract class AbstractReactorConnectionOperations<C> implements ReactorC
     }
 
     private <T> Mono<T> existingConnectionMono(ConnectionDefinition definition, Function<ConnectionStatus<C>, Mono<T>> callback, C clientSession) {
-        return callback.apply(new DefaultReactiveConnectionStatus<>(clientSession, definition, false));
+        return applyCallbackMono(callback, new DefaultReactiveConnectionStatus<>(clientSession, definition, false));
     }
 
     private <T> Mono<T> openConnectionMono(ConnectionDefinition definition, Function<ConnectionStatus<C>, Mono<T>> callback) {
         return Mono.usingWhen(
             Mono.from(openConnection(definition)).map(connection -> new DefaultReactiveConnectionStatus<>(connection, definition, true)),
-            connectionStatus -> callback.apply(connectionStatus).contextWrite(ctx -> addClientSession(ctx, connectionStatus)),
+            connectionStatus -> applyCallbackMono(callback, connectionStatus).contextWrite(ctx -> addClientSession(ctx, connectionStatus)),
             connectionStatus -> connectionStatus.onComplete(() -> closeConnection(connectionStatus.getConnection(), definition)),
             (connectionStatus, throwable) -> connectionStatus.onError(throwable, () -> closeConnection(connectionStatus.getConnection(), definition)),
             connectionStatus -> connectionStatus.onCancel(() -> closeConnection(connectionStatus.getConnection(), definition))
@@ -164,6 +164,22 @@ public abstract class AbstractReactorConnectionOperations<C> implements ReactorC
         return findConnectionStatus(contextView)
             .map(ConnectionStatus::getConnection)
             .orElse(null);
+    }
+
+    private <T> Flux<T> applyCallbackFlux(Function<ConnectionStatus<C>, Flux<T>> callback, DefaultReactiveConnectionStatus<C> connectionStatus) {
+        try {
+            return callback.apply(connectionStatus);
+        } catch (Exception e) {
+            return Flux.error(e);
+        }
+    }
+
+    private <T> Mono<T> applyCallbackMono(Function<ConnectionStatus<C>, Mono<T>> callback, DefaultReactiveConnectionStatus<C> connectionStatus) {
+        try {
+            return callback.apply(connectionStatus);
+        } catch (Exception e) {
+            return Mono.error(e);
+        }
     }
 
     private record ClientSessionPropagatedContext<C>(

--- a/data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/ColumnNameExistenceAwareResultSetReader.java
+++ b/data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/ColumnNameExistenceAwareResultSetReader.java
@@ -1,0 +1,66 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.jdbc.mapper;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.data.model.DataType;
+import io.micronaut.data.runtime.mapper.AbstractDelegatingResultReader;
+
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.Set;
+
+/**
+ * The reader that will return null if the column doesn't exist in the result.
+ *
+ * @author Denis Stepanov
+ * @since 4.9
+ */
+@Internal
+public class ColumnNameExistenceAwareResultSetReader extends AbstractDelegatingResultReader<ResultSet, String> {
+
+    private Set<String> knownColumns;
+
+    public ColumnNameExistenceAwareResultSetReader() {
+        super(new ColumnNameResultSetReader());
+    }
+
+    @Override
+    public Object readDynamic(ResultSet resultSet, String index, DataType dataType) {
+        if (!containsColumnName(resultSet, index)) {
+            return null;
+        }
+        return super.readDynamic(resultSet, index, dataType);
+    }
+
+    private boolean containsColumnName(ResultSet resultSet, String name) {
+        if (knownColumns == null) {
+            try {
+                ResultSetMetaData rsmd = resultSet.getMetaData();
+                int columnsCount = rsmd.getColumnCount();
+                knownColumns = CollectionUtils.newHashSet(columnsCount);
+                for (int x = 1; x <= columnsCount; x++) {
+                    knownColumns.add(rsmd.getColumnLabel(x).toLowerCase());
+                }
+            } catch (SQLException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return knownColumns.contains(name);
+    }
+}

--- a/data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java
+++ b/data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java
@@ -31,13 +31,12 @@ import io.micronaut.core.util.ArgumentUtils;
 import io.micronaut.core.util.CollectionUtils;
 import io.micronaut.data.connection.ConnectionOperations;
 import io.micronaut.data.connection.annotation.Connectable;
-import io.micronaut.data.connection.jdbc.advice.DelegatingDataSource;
 import io.micronaut.data.exceptions.DataAccessException;
 import io.micronaut.data.jdbc.config.DataJdbcConfiguration;
 import io.micronaut.data.jdbc.convert.JdbcConversionContext;
 import io.micronaut.data.jdbc.mapper.ColumnIndexCallableResultReader;
 import io.micronaut.data.jdbc.mapper.ColumnIndexResultSetReader;
-import io.micronaut.data.jdbc.mapper.ColumnNameCallableResultReader;
+import io.micronaut.data.jdbc.mapper.ColumnNameExistenceAwareResultSetReader;
 import io.micronaut.data.jdbc.mapper.ColumnNameResultSetReader;
 import io.micronaut.data.jdbc.mapper.JdbcQueryStatement;
 import io.micronaut.data.jdbc.mapper.SqlResultConsumer;
@@ -148,7 +147,6 @@ public final class DefaultJdbcRepositoryOperations extends AbstractSqlRepository
     private final ConnectionOperations<Connection> connectionOperations;
     private final TransactionOperations<Connection> transactionOperations;
     private final DataSource dataSource;
-    private final DataSource unwrapedDataSource;
     private ExecutorAsyncOperations asyncOperations;
     private ExecutorService executorService;
     private final SyncCascadeOperations<JdbcOperationContext> cascadeOperations;
@@ -157,7 +155,6 @@ public final class DefaultJdbcRepositoryOperations extends AbstractSqlRepository
     private final SchemaTenantResolver schemaTenantResolver;
     private final JdbcSchemaHandler schemaHandler;
 
-    private final ColumnNameCallableResultReader columnNameCallableResultReader;
     private final ColumnIndexCallableResultReader columnIndexCallableResultReader;
     private final Map<Dialect, List<SqlExceptionMapper>> sqlExceptionMappers = new EnumMap<>(Dialect.class);
 
@@ -217,12 +214,10 @@ public final class DefaultJdbcRepositoryOperations extends AbstractSqlRepository
         ArgumentUtils.requireNonNull("dataSource", dataSource);
         ArgumentUtils.requireNonNull("transactionOperations", transactionOperations);
         this.dataSource = dataSource;
-        this.unwrapedDataSource = DelegatingDataSource.unwrapDataSource(dataSource);
         this.transactionOperations = transactionOperations;
         this.executorService = executorService;
         this.cascadeOperations = new SyncCascadeOperations<>(conversionService, this);
         this.jdbcConfiguration = jdbcConfiguration;
-        this.columnNameCallableResultReader = new ColumnNameCallableResultReader(conversionService);
         this.columnIndexCallableResultReader = new ColumnIndexCallableResultReader(conversionService);
         if (CollectionUtils.isNotEmpty(sqlExceptionMapperList)) {
             for (SqlExceptionMapper sqlExceptionMapper : sqlExceptionMapperList) {
@@ -235,6 +230,11 @@ public final class DefaultJdbcRepositoryOperations extends AbstractSqlRepository
                 sqlExceptionMappers.put(dialect, dialectSqlExceptionMapperList);
             }
         }
+    }
+
+    @Override
+    protected ResultReader<ResultSet, String> createColumnNameResultSetReaderWithColumnExistenceAware() {
+        return new ColumnNameExistenceAwareResultSetReader();
     }
 
     @NonNull

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2RepositorySpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2RepositorySpec.groovy
@@ -41,7 +41,10 @@ import io.micronaut.data.tck.repositories.TimezoneBasicTypesRepository
 import io.micronaut.data.tck.repositories.UserRepository
 import io.micronaut.data.tck.repositories.UserRoleRepository
 import io.micronaut.data.tck.tests.AbstractRepositorySpec
+import spock.lang.PendingFeature
 import spock.lang.Shared
+
+import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.nameEqualsCaseInsensitive
 
 class H2RepositorySpec extends AbstractRepositorySpec implements H2TestPropertyProvider {
 
@@ -243,6 +246,15 @@ class H2RepositorySpec extends AbstractRepositorySpec implements H2TestPropertyP
     @Override
     protected boolean skipQueryByDataArray() {
         return true
+    }
+
+    @PendingFeature
+    void "test criteria lower select" () {
+        when:
+            savePersons(["Jeff", "James"])
+            def person = personRepository.findOne(nameEqualsCaseInsensitive("james"))
+        then:
+            person.isPresent()
     }
 
     void "test manual joining on many ended association"() {

--- a/data-r2dbc/build.gradle
+++ b/data-r2dbc/build.gradle
@@ -55,7 +55,7 @@ dependencies {
     testImplementation mnSql.mariadb.java.client
 
     // mysql
-    testImplementation mnR2dbc.r2dbc.mysql
+    testImplementation mnR2dbc.r2dbc.io.asyncer.mysql
     testImplementation mnSql.mysql.connector.java
 
     // h2

--- a/data-r2dbc/src/main/java/io/micronaut/data/r2dbc/mapper/ColumnNameExistenceAwareR2dbcResultSetReader.java
+++ b/data-r2dbc/src/main/java/io/micronaut/data/r2dbc/mapper/ColumnNameExistenceAwareR2dbcResultSetReader.java
@@ -1,0 +1,63 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.r2dbc.mapper;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.data.model.DataType;
+import io.micronaut.data.runtime.mapper.AbstractDelegatingResultReader;
+import io.r2dbc.spi.ColumnMetadata;
+import io.r2dbc.spi.Row;
+import io.r2dbc.spi.RowMetadata;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * The reader that will return null if the column doesn't exist in the result.
+ *
+ * @author Denis Stepanov
+ * @since 4.9
+ */
+@Internal
+public class ColumnNameExistenceAwareR2dbcResultSetReader extends AbstractDelegatingResultReader<Row, String> {
+
+    private Set<String> knownColumns;
+
+    public ColumnNameExistenceAwareR2dbcResultSetReader() {
+        super(new ColumnNameR2dbcResultReader());
+    }
+
+    @Override
+    public Object readDynamic(Row row, String index, DataType dataType) {
+        if (!containsColumnName(row, index)) {
+            return null;
+        }
+        return super.readDynamic(row, index, dataType);
+    }
+
+    private boolean containsColumnName(Row row, String name) {
+        if (knownColumns == null) {
+            RowMetadata metadata = row.getMetadata();
+            List<? extends ColumnMetadata> columnMetadatas = metadata.getColumnMetadatas();
+            knownColumns = CollectionUtils.newHashSet(columnMetadatas.size());
+            for (ColumnMetadata columnMetadata : columnMetadatas) {
+                knownColumns.add(columnMetadata.getName().toLowerCase());
+            }
+        }
+        return knownColumns.contains(name.toLowerCase());
+    }
+}

--- a/data-r2dbc/src/main/java/io/micronaut/data/r2dbc/operations/DefaultR2dbcRepositoryOperations.java
+++ b/data-r2dbc/src/main/java/io/micronaut/data/r2dbc/operations/DefaultR2dbcRepositoryOperations.java
@@ -62,12 +62,14 @@ import io.micronaut.data.r2dbc.annotation.R2dbcRepository;
 import io.micronaut.data.r2dbc.config.DataR2dbcConfiguration;
 import io.micronaut.data.r2dbc.convert.R2dbcConversionContext;
 import io.micronaut.data.r2dbc.mapper.ColumnIndexR2dbcResultReader;
+import io.micronaut.data.r2dbc.mapper.ColumnNameExistenceAwareR2dbcResultSetReader;
 import io.micronaut.data.r2dbc.mapper.ColumnNameR2dbcResultReader;
 import io.micronaut.data.r2dbc.mapper.R2dbcQueryStatement;
 import io.micronaut.data.r2dbc.transaction.R2dbcReactorTransactionOperations;
 import io.micronaut.data.runtime.convert.DataConversionService;
 import io.micronaut.data.runtime.convert.RuntimePersistentPropertyConversionContext;
 import io.micronaut.data.runtime.date.DateTimeProvider;
+import io.micronaut.data.runtime.mapper.ResultReader;
 import io.micronaut.data.runtime.mapper.sql.SqlResultEntityTypeMapper;
 import io.micronaut.data.runtime.mapper.sql.SqlTypeMapper;
 import io.micronaut.data.runtime.multitenancy.SchemaTenantResolver;
@@ -226,6 +228,11 @@ final class DefaultR2dbcRepositoryOperations extends AbstractSqlRepositoryOperat
                 r2dbcExceptionMappers.put(dialect, dialectR2dbcExceptionMapperList);
             }
         }
+    }
+
+    @Override
+    protected ResultReader<Row, String> createColumnNameResultSetReaderWithColumnExistenceAware() {
+        return new ColumnNameExistenceAwareR2dbcResultSetReader();
     }
 
     @Override

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/criteria/AbstractSpecificationInterceptor.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/criteria/AbstractSpecificationInterceptor.java
@@ -31,6 +31,7 @@ import io.micronaut.data.model.Pageable;
 import io.micronaut.data.model.Pageable.Mode;
 import io.micronaut.data.model.Sort;
 import io.micronaut.data.model.jpa.criteria.PersistentEntityFrom;
+import io.micronaut.data.model.jpa.criteria.impl.AbstractPersistentEntityCriteriaQuery;
 import io.micronaut.data.model.jpa.criteria.impl.QueryResultPersistentEntityCriteriaQuery;
 import io.micronaut.data.model.query.JoinPath;
 import io.micronaut.data.model.query.builder.QueryBuilder;
@@ -62,6 +63,7 @@ import jakarta.persistence.criteria.Order;
 import jakarta.persistence.criteria.Path;
 import jakarta.persistence.criteria.Predicate;
 import jakarta.persistence.criteria.Root;
+import jakarta.persistence.criteria.Selection;
 
 import java.util.ArrayList;
 import java.util.Collection;
@@ -326,13 +328,15 @@ public abstract class AbstractSpecificationInterceptor<T, R> extends AbstractQue
         QueryResult queryResult = ((QueryResultPersistentEntityCriteriaQuery) criteriaQuery).buildQuery(context, sqlQueryBuilder);
         Set<JoinPath> joinPaths = mergeJoinPaths(methodJoinPaths, queryResult.getJoinPaths());
         Class<E> rootEntity = getRequiredRootEntity(context);
+        Selection<?> selection = ((AbstractPersistentEntityCriteriaQuery<?>) criteriaQuery).getSelection();
+        boolean isCompoundSelection = selection != null && selection.isCompoundSelection();
         if (type == Type.FIND_ONE) {
             return QueryResultStoredQuery.single(OperationType.QUERY, context.getName(), context.getAnnotationMetadata(),
-                queryResult, rootEntity, criteriaQuery.getResultType(), joinPaths);
+                queryResult, rootEntity, criteriaQuery.getResultType(), isCompoundSelection, joinPaths);
         }
         Pageable pageable = findPageable(context);
         return QueryResultStoredQuery.many(context.getName(), context.getAnnotationMetadata(), queryResult, rootEntity,
-            criteriaQuery.getResultType(), !pageable.isUnpaged(), joinPaths);
+            criteriaQuery.getResultType(), !pageable.isUnpaged(), isCompoundSelection, joinPaths);
     }
 
     private <N> CriteriaQuery<N> buildInternalQuery(MethodInvocationContext<T, R> context, Type type, Set<JoinPath> methodJoinPaths) {

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/mapper/AbstractDelegatingResultReader.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/mapper/AbstractDelegatingResultReader.java
@@ -1,0 +1,151 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.runtime.mapper;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.type.Argument;
+import io.micronaut.data.exceptions.DataAccessException;
+import io.micronaut.data.model.DataType;
+
+import java.math.BigDecimal;
+import java.sql.Time;
+import java.util.Date;
+import java.util.UUID;
+
+/**
+ * The delegating {@link ResultReader}.
+ *
+ * @param <RS> The result set
+ * @param <ID> The index type
+ * @author Denis Stepanov
+ * @since 4.9
+ */
+@Internal
+public abstract class AbstractDelegatingResultReader<RS, ID> implements ResultReader<RS, ID> {
+
+    protected final ResultReader<RS, ID> delegate;
+
+    protected AbstractDelegatingResultReader(ResultReader<RS, ID> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public <T> T convertRequired(Object value, Class<T> type) {
+        return delegate.convertRequired(value, type);
+    }
+
+    @Override
+    public <T> T convertRequired(Object value, Argument<T> type) {
+        return delegate.convertRequired(value, type);
+    }
+
+    @Override
+    public <T> T getRequiredValue(RS resultSet, ID name, Class<T> type) throws DataAccessException {
+        return delegate.getRequiredValue(resultSet, name, type);
+    }
+
+    @Override
+    public boolean next(RS resultSet) {
+        return delegate.next(resultSet);
+    }
+
+    @Override
+    public Object readDynamic(RS resultSet, ID index, DataType dataType) {
+        return delegate.readDynamic(resultSet, index, dataType);
+    }
+
+    @Override
+    public long readLong(RS resultSet, ID name) {
+        return delegate.readLong(resultSet, name);
+    }
+
+    @Override
+    public char readChar(RS resultSet, ID name) {
+        return delegate.readChar(resultSet, name);
+    }
+
+    @Override
+    public Date readDate(RS resultSet, ID name) {
+        return delegate.readDate(resultSet, name);
+    }
+
+    @Override
+    public Date readTimestamp(RS resultSet, ID index) {
+        return delegate.readTimestamp(resultSet, index);
+    }
+
+    @Override
+    public Time readTime(RS resultSet, ID index) {
+        return delegate.readTime(resultSet, index);
+    }
+
+    @Override
+    public String readString(RS resultSet, ID name) {
+        return delegate.readString(resultSet, name);
+    }
+
+    @Override
+    public UUID readUUID(RS resultSet, ID name) {
+        return delegate.readUUID(resultSet, name);
+    }
+
+    @Override
+    public int readInt(RS resultSet, ID name) {
+        return delegate.readInt(resultSet, name);
+    }
+
+    @Override
+    public boolean readBoolean(RS resultSet, ID name) {
+        return delegate.readBoolean(resultSet, name);
+    }
+
+    @Override
+    public float readFloat(RS resultSet, ID name) {
+        return delegate.readFloat(resultSet, name);
+    }
+
+    @Override
+    public byte readByte(RS resultSet, ID name) {
+        return delegate.readByte(resultSet, name);
+    }
+
+    @Override
+    public short readShort(RS resultSet, ID name) {
+        return delegate.readShort(resultSet, name);
+    }
+
+    @Override
+    public double readDouble(RS resultSet, ID name) {
+        return delegate.readDouble(resultSet, name);
+    }
+
+    @Override
+    public BigDecimal readBigDecimal(RS resultSet, ID name) {
+        return delegate.readBigDecimal(resultSet, name);
+    }
+
+    @Override
+    public byte[] readBytes(RS resultSet, ID name) {
+        return delegate.readBytes(resultSet, name);
+    }
+
+    @Override
+    public ConversionService getConversionService() {
+        return delegate.getConversionService();
+    }
+
+}

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/AbstractSqlRepositoryOperations.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/AbstractSqlRepositoryOperations.java
@@ -166,6 +166,13 @@ public abstract class AbstractSqlRepositoryOperations<RS, PS, Exc extends Except
         }
     }
 
+    /**
+     * @return The result reader that will check for the column existence and return null for {@link ResultReader#readDynamic(Object, Object, DataType)}
+     */
+    protected ResultReader<RS, String> createColumnNameResultSetReaderWithColumnExistenceAware() {
+        return columnNameResultSetReader;
+    }
+
     @Override
     public <E, R> PreparedQuery<E, R> decorate(PreparedQuery<E, R> preparedQuery) {
         return new DefaultSqlPreparedQuery<>(preparedQuery);
@@ -677,9 +684,11 @@ public abstract class AbstractSqlRepositoryOperations<RS, PS, Exc extends Except
             return createQueryResultMapper(preparedQuery, column, jsonDataType, rsType, persistentEntity, loadListener);
         }
         if (isEntityResult) {
+            ResultReader<RS, String> resultReader =
+                preparedQuery.isDtoProjection() ? createColumnNameResultSetReaderWithColumnExistenceAware() : columnNameResultSetReader;
             return new SqlResultEntityTypeMapper<>(
                 getEntity(preparedQuery.getResultType()),
-                columnNameResultSetReader,
+                resultReader,
                 preparedQuery.getJoinPaths(),
                 sqlJsonColumnMapperProvider.getJsonColumnReader(preparedQuery, rsType),
                 loadListener,

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/query/internal/BasicStoredQuery.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/query/internal/BasicStoredQuery.java
@@ -75,6 +75,24 @@ public class BasicStoredQuery<E, R> implements StoredQuery<E, R> {
                             boolean isSingleResult,
                             boolean isCount,
                             OperationType operationType) {
+        this(name, annotationMetadata, query, expandableQueryParts, queryParameterBindings,
+            rootEntity, resultType, pageable, isSingleResult, isCount,
+            rootEntity != resultType && (DataType.forType(resultType) == DataType.OBJECT && BeanIntrospector.SHARED.findIntrospection(resultType).isPresent()),
+            operationType);
+    }
+
+    public BasicStoredQuery(String name,
+                            AnnotationMetadata annotationMetadata,
+                            String query,
+                            String[] expandableQueryParts,
+                            List<QueryParameterBinding> queryParameterBindings,
+                            Class<E> rootEntity,
+                            Class<R> resultType,
+                            boolean pageable,
+                            boolean isSingleResult,
+                            boolean isCount,
+                            boolean isDto,
+                            OperationType operationType) {
         this.name = name;
         this.annotationMetadata = annotationMetadata;
         this.query = query;
@@ -88,7 +106,7 @@ public class BasicStoredQuery<E, R> implements StoredQuery<E, R> {
         this.operationType = operationType;
         this.resultDataType = isCount ? DataType.forType(resultType) : (rootEntity == resultType) ? DataType.ENTITY : DataType.forType(resultType);
         this.rawQuery = annotationMetadata.stringValue(Query.class, DataMethod.META_MEMBER_RAW_QUERY).isPresent();
-        this.isDto = resultDataType == DataType.OBJECT && BeanIntrospector.SHARED.findIntrospection(resultType).isPresent();
+        this.isDto = isDto;
     }
 
     @Override

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/query/internal/QueryResultStoredQuery.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/query/internal/QueryResultStoredQuery.java
@@ -56,17 +56,45 @@ public final class QueryResultStoredQuery<E, R> extends BasicStoredQuery<E, R> {
                                   OperationType operationType,
                                   Collection<JoinPath> joinPaths) {
         super(name,
-                annotationMetadata,
-                queryResult.getQuery(),
-                queryResult.getParameterBindings().stream()
-                        .anyMatch(io.micronaut.data.model.query.builder.QueryParameterBinding::isExpandable) ? queryResult.getQueryParts().toArray(new String[0]) : null,
-                map(queryResult.getParameterBindings()),
-                rootEntity,
-                resultType,
-                pageable,
-                isSingleResult,
-                isCount,
-                operationType);
+            annotationMetadata,
+            queryResult.getQuery(),
+            queryResult.getParameterBindings().stream()
+                .anyMatch(io.micronaut.data.model.query.builder.QueryParameterBinding::isExpandable) ? queryResult.getQueryParts().toArray(new String[0]) : null,
+            map(queryResult.getParameterBindings()),
+            rootEntity,
+            resultType,
+            pageable,
+            isSingleResult,
+            isCount,
+            operationType);
+        this.queryResult = queryResult;
+        this.joinPaths = joinPaths == null ? Collections.emptySet() : Set.copyOf(joinPaths);
+    }
+
+    public QueryResultStoredQuery(String name,
+                                  AnnotationMetadata annotationMetadata,
+                                  QueryResult queryResult,
+                                  Class<E> rootEntity,
+                                  Class<R> resultType,
+                                  boolean pageable,
+                                  boolean isSingleResult,
+                                  boolean isCount,
+                                  boolean isDto,
+                                  OperationType operationType,
+                                  Collection<JoinPath> joinPaths) {
+        super(name,
+            annotationMetadata,
+            queryResult.getQuery(),
+            queryResult.getParameterBindings().stream()
+                .anyMatch(io.micronaut.data.model.query.builder.QueryParameterBinding::isExpandable) ? queryResult.getQueryParts().toArray(new String[0]) : null,
+            map(queryResult.getParameterBindings()),
+            rootEntity,
+            resultType,
+            pageable,
+            isSingleResult,
+            isCount,
+            isDto,
+            operationType);
         this.queryResult = queryResult;
         this.joinPaths = joinPaths == null ? Collections.emptySet() : Set.copyOf(joinPaths);
     }
@@ -80,20 +108,31 @@ public final class QueryResultStoredQuery<E, R> extends BasicStoredQuery<E, R> {
     }
 
     public static <T, R> QueryResultStoredQuery<T, R> single(OperationType operationType,
-                                                          String name,
-                                                          AnnotationMetadata annotationMetadata,
-                                                          QueryResult queryResult,
-                                                          Class<T> rootEntity,
-                                                          Class<R> resultType,
-                                                          Collection<JoinPath> joinPaths) {
+                                                             String name,
+                                                             AnnotationMetadata annotationMetadata,
+                                                             QueryResult queryResult,
+                                                             Class<T> rootEntity,
+                                                             Class<R> resultType,
+                                                             Collection<JoinPath> joinPaths) {
         return new QueryResultStoredQuery<>(name, annotationMetadata, queryResult, rootEntity, resultType == Object.class ? (Class<R>) rootEntity : resultType, false, true, false, operationType, joinPaths);
     }
 
+    public static <T, R> QueryResultStoredQuery<T, R> single(OperationType operationType,
+                                                             String name,
+                                                             AnnotationMetadata annotationMetadata,
+                                                             QueryResult queryResult,
+                                                             Class<T> rootEntity,
+                                                             Class<R> resultType,
+                                                             boolean isDto,
+                                                             Collection<JoinPath> joinPaths) {
+        return new QueryResultStoredQuery<>(name, annotationMetadata, queryResult, rootEntity, resultType == Object.class ? (Class<R>) rootEntity : resultType, false, true, false, isDto, operationType, joinPaths);
+    }
+
     public static <T> QueryResultStoredQuery<T, T> many(String name,
-                                                           AnnotationMetadata annotationMetadata,
-                                                           QueryResult queryResult,
-                                                           Class<T> rootEntity,
-                                                           boolean pageable) {
+                                                        AnnotationMetadata annotationMetadata,
+                                                        QueryResult queryResult,
+                                                        Class<T> rootEntity,
+                                                        boolean pageable) {
         return new QueryResultStoredQuery<>(name, annotationMetadata, queryResult, rootEntity, rootEntity, pageable, false, false, OperationType.QUERY, Collections.emptySet());
     }
 
@@ -107,6 +146,17 @@ public final class QueryResultStoredQuery<E, R> extends BasicStoredQuery<E, R> {
         return new QueryResultStoredQuery<>(name, annotationMetadata, queryResult, rootEntity, resultType == Object.class ? (Class<R>) rootEntity : resultType, pageable, false, false, OperationType.QUERY, joinPaths);
     }
 
+    public static <T, R> QueryResultStoredQuery<T, R> many(String name,
+                                                           AnnotationMetadata annotationMetadata,
+                                                           QueryResult queryResult,
+                                                           Class<T> rootEntity,
+                                                           Class<R> resultType,
+                                                           boolean pageable,
+                                                           boolean isDto,
+                                                           Collection<JoinPath> joinPaths) {
+        return new QueryResultStoredQuery<>(name, annotationMetadata, queryResult, rootEntity, resultType == Object.class ? (Class<R>) rootEntity : resultType, pageable, false, false, isDto, OperationType.QUERY, joinPaths);
+    }
+
     public static <T> QueryResultStoredQuery<T, Long> count(String name,
                                                             AnnotationMetadata annotationMetadata,
                                                             QueryResult queryResult,
@@ -118,7 +168,7 @@ public final class QueryResultStoredQuery<E, R> extends BasicStoredQuery<E, R> {
         List<QueryParameterBinding> queryParameters = new ArrayList<>(parameterBindings.size());
         for (io.micronaut.data.model.query.builder.QueryParameterBinding p : parameterBindings) {
             queryParameters.add(
-                    new QueryResultParameterBinding(p, queryParameters)
+                new QueryResultParameterBinding(p, queryParameters)
             );
         }
         return queryParameters;

--- a/data-tck/src/main/groovy/io/micronaut/data/tck/tests/AbstractRepositorySpec.groovy
+++ b/data-tck/src/main/groovy/io/micronaut/data/tck/tests/AbstractRepositorySpec.groovy
@@ -51,6 +51,8 @@ import io.micronaut.data.tck.entities.Meal
 import io.micronaut.data.tck.entities.Nose
 import io.micronaut.data.tck.entities.Page
 import io.micronaut.data.tck.entities.Person
+import io.micronaut.data.tck.entities.PersonDto
+import io.micronaut.data.tck.entities.PersonDto2
 import io.micronaut.data.tck.entities.Student
 import io.micronaut.data.tck.entities.TimezoneBasicTypes
 import io.micronaut.data.tck.jdbc.entities.Role
@@ -86,6 +88,8 @@ import static io.micronaut.data.tck.repositories.BookSpecifications.titleEqualsW
 import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.distinct
 import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.idsIn
 import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.nameEquals
+import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.nameEqualsCaseInsensitive
+import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.personWithOnlyNameAndAgeByName
 import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.setIncome
 import static io.micronaut.data.tck.repositories.PersonRepository.Specifications.setName
 
@@ -2395,7 +2399,58 @@ abstract class AbstractRepositorySpec extends Specification {
             !existsNotPredicateSpec
             existsQuerySpec
             !existsNotQuerySpec
+    }
 
+    void "test criteria select" () {
+        when:
+            personRepository.save(new Person(name: "Denis", age: 123, income: 10000, enabled: false))
+            def person = personRepository.findOne(personWithOnlyNameAndAgeByName("Denis")).get()
+        then:
+            person.id == null
+            person.income == null
+            person.name == "Denis"
+            person.age == 123
+            !person.income
+            person.enabled
+    }
+
+    void "test criteria DTO projection"() {
+        when:
+            personRepository.deleteAll()
+            personRepository.save(new Person(name: "Fred1", age: 50))
+            personRepository.save(new Person(name: "Fred2", age: 18))
+        then:
+            def dto = personRepository.findOne(new CriteriaQueryBuilder<PersonDto>() {
+                @Override
+                CriteriaQuery<PersonDto> build(CriteriaBuilder criteriaBuilder) {
+                    def query = criteriaBuilder.createQuery(PersonDto)
+                    def root = query.from(Person)
+                    query.multiselect(root.<Integer>get("age"))
+                    query.where(criteriaBuilder.equal(root.<String>get("name"), "Fred1"))
+                    return query
+                }
+            })
+            dto.age == 50
+    }
+
+    void "test criteria DTO projection 2"() {
+        when:
+            personRepository.deleteAll()
+            personRepository.save(new Person(name: "Fred1", age: 50))
+            personRepository.save(new Person(name: "Fred2", age: 18))
+        then:
+            def dto = personRepository.findOne(new CriteriaQueryBuilder<PersonDto2>() {
+                @Override
+                CriteriaQuery<PersonDto2> build(CriteriaBuilder criteriaBuilder) {
+                    def query = criteriaBuilder.createQuery(PersonDto2)
+                    def root = query.from(Person)
+                    query.multiselect(root.<String>get("name"), root.<Integer>get("age"))
+                    query.where(criteriaBuilder.equal(root.<String>get("name"), "Fred1"))
+                    return query
+                }
+            })
+            dto.name() == "Fred1"
+            dto.age() == 50
     }
 
     void "test join/fetch"() {

--- a/data-tck/src/main/java/io/micronaut/data/tck/entities/Person.java
+++ b/data-tck/src/main/java/io/micronaut/data/tck/entities/Person.java
@@ -25,6 +25,7 @@ import jakarta.persistence.Id;
 public class Person {
     @Id
     @GeneratedValue
+    @Nullable
     private Long id;
 
     private String name;

--- a/data-tck/src/main/java/io/micronaut/data/tck/entities/PersonDto2.java
+++ b/data-tck/src/main/java/io/micronaut/data/tck/entities/PersonDto2.java
@@ -1,0 +1,22 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.tck.entities;
+
+import io.micronaut.core.annotation.Introspected;
+
+@Introspected
+public record PersonDto2(String name, int age) {
+}

--- a/data-tck/src/main/java/io/micronaut/data/tck/repositories/PersonRepository.java
+++ b/data-tck/src/main/java/io/micronaut/data/tck/repositories/PersonRepository.java
@@ -36,6 +36,10 @@ import io.micronaut.data.repository.jpa.criteria.UpdateSpecification;
 import io.micronaut.data.tck.entities.Person;
 import io.micronaut.data.tck.entities.TotalDto;
 import io.reactivex.Single;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
 
 import java.util.Arrays;
 import java.util.List;
@@ -220,6 +224,13 @@ public interface PersonRepository extends CrudRepository<Person, Long>, Pageable
             return (root, query, criteriaBuilder) -> {
                 query.set("name", name == null ? criteriaBuilder.nullLiteral(String.class) : name);
                 return null;
+            };
+        }
+
+        public static QuerySpecification<Person> personWithOnlyNameAndAgeByName(String name) {
+            return (root, query, criteriaBuilder) -> {
+                query.multiselect(root.get("name").alias("name"), root.get("age").alias("age"));
+                return criteriaBuilder.equal(root.get("name"), name);
             };
         }
     }

